File: Code/Chapter04/AccountStatus/c#/AccountStatus/AccountStatus.cs
/*
 AccountStatus.cs

  Created by Ray Rischpater on 12 July 2003
  Copyright (c) 2003 by the author. All rights reserved.
  This file is provided without any warranty, express or 
  implied, including but not limited to fitness of purpose.
  You may use this file so long as the copyright attribution
  remains.

 This file is part of the book "Developing Applications with 
 eBay" written by Ray Rischpater and available from 
 Apress, LP.
   
 Dependencies
  - eBay SDK
  - Microsoft .NET
  
 Description
  This package provides an application that demonstrates 
  eBay.SDK.Model.Account classes and eBay.SDK.API.GetAccount.
*/

using System;
using System.IO;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using eBay.SDK.Model.Account;
using eBay.SDK.API;

namespace com.lothlorien.ebaysdkbook
{
	/// <summary>
	/// Summary description for Class1.
	/// </summary>
	public class AccountStatus  : com.lothlorien.ebaysdkbook.eBaySampleApplication
	{
		private System.Windows.Forms.Panel panel1;
		private System.Windows.Forms.ListBox viewtype;
		private System.Windows.Forms.ListBox pagenumber;
		private System.Windows.Forms.DateTimePicker startdate;
		private System.Windows.Forms.DateTimePicker enddate;

		private System.Windows.Forms.Label label1;
		private System.Windows.Forms.Label label2;
		private System.Windows.Forms.Label label3;
		private System.Windows.Forms.TextBox userid;
		private System.Windows.Forms.Button findButton;
		private System.Windows.Forms.Panel panel2;
		private System.Windows.Forms.RichTextBox result;
		
		public AccountStatus()
		{
			InitializeComponent();
			// A reasonable default user
			this.userid.Text = this.ApiSession.RequestUserId;
			this.Gui = this.panel1;
		}

		public int Month
		{
			get
			{
				return startdate.Value.Month;
			}
		}

		public int Year
		{
			get
			{
				return startdate.Value.Year;
			}
		}

		private void find_Click(object sender, System.EventArgs e)
		{
			IGetAccountCall api = new GetAccountCall();
			IAccount theAccount = null;
			string resultString = null;

			// We're busy.
			this.Busy = true;
			this.StatusText = "Downloading...";

			// Set the API for this call.
			api.CompatibilityLevel = 319;
			api.ErrorLevel = 
				ErrorLevelEnum.BothShortAndLongErrorStrings;
			api.ApiCallSession = this.ApiSession;
			api.PageNumber = System.Convert.ToInt32( this.pagenumber.SelectedItem );
			api.ViewType = (AccountViewEnum)this.viewtype.SelectedIndex;
			
			switch( api.ViewType )
			{
				case AccountViewEnum.ViewByPeriod:
					api.PeriodViewOption = new PeriodViewSettings();
					api.PeriodViewOption.DateRange = new DateRangeImpl();
					api.PeriodViewOption.DateRange.BeginDate = this.startdate.Value;
					api.PeriodViewOption.DateRange.EndDate = this.enddate.Value;
					break;

				case AccountViewEnum.ViewByInvoice:
					api.InvoiceViewOption = new InvoiceViewSettings();
					api.InvoiceViewOption.InvoiceMonth = (InvoiceMonthEnum)this.Month;
					api.InvoiceViewOption.InvoiceYear = this.Year;
					break;
			}

			try
			{
				theAccount = api.GetAccount();
			}
			catch( Exception ex )
			{
				resultString =  "**Error**\n" + ex.Message;
			}

			// Create a nice string with the results.
			if ( theAccount != null )
			{
				resultString = "Account: " + theAccount.Id + "\n";
				// Summary of invoice view.
				if ( theAccount.InvoiceView != null )
				{
					resultString += "Amount Due:\t" + theAccount.InvoiceView.InvoiceBalance + "\n";
					resultString += "Invoice Date:\t" + theAccount.InvoiceView.InvoiceDate.ToShortDateString() + "\n";
				}
				// Summary of period view.
				if ( theAccount.PeriodView != null )
				{
					resultString += "Last Invoice Date:\t\t" + theAccount.PeriodView.LastInvoiceDate.ToString() + "\n";
					resultString += "Last Invoice Amount:\t" + theAccount.PeriodView.LastInvoiceAmount.ToString() + "\n";
					resultString += "Last Payment Date:\t\t" + theAccount.PeriodView.LastPaymentDate + "\n";
					resultString += "Last Payment Amount:\t" + theAccount.PeriodView.LastAmountPaid + "\n";
					resultString += "Current Balance:\t\t" + theAccount.PeriodView.CurrentBalance + "\n";
				}
				// Enumerate each of the account activities
				resultString += "\nActivity History\n";
				foreach ( AccountActivity activity in theAccount.Activities )
				{
					resultString += activity.Date.ToShortDateString() + "\t";
					resultString += activity.Id.ToString() + "\t" + activity.ItemId + " " + activity.Memo + "\t";
					resultString += activity.Credit != 0 ? activity.Credit.ToString() : "";
					resultString += activity.Debit != 0 ? "(" + activity.Debit.ToString() + ")" : "";
					resultString += "\n";
				}
			}

			// Update the pagenumber menu with the list pages
			this.pagenumber.Items.Clear();
			for ( int i = 0; i <= api.TotalPages; i++ )
			{
				this.pagenumber.Items.Add( (i+1).ToString() );
			}
			this.pagenumber.SelectedIndex = 0;

			// Update the UI.
			this.result.SuspendLayout();
			result.Text = resultString ;
			this.result.Enabled = true;
			this.result.ResumeLayout();

			// We're no longer busy.
			this.StatusText = "";
			this.Busy = false;
		}


		private void InitializeComponent()
		{
			this.panel1 = new System.Windows.Forms.Panel();
			this.panel2 = new System.Windows.Forms.Panel();
			this.result = new System.Windows.Forms.RichTextBox();
			this.findButton = new System.Windows.Forms.Button();
			this.userid = new System.Windows.Forms.TextBox();
			this.label1 = new System.Windows.Forms.Label();
			this.label2 = new System.Windows.Forms.Label();
			this.label3 = new System.Windows.Forms.Label();
			this.viewtype = new System.Windows.Forms.ListBox();
			this.pagenumber = new System.Windows.Forms.ListBox();
			this.startdate = new System.Windows.Forms.DateTimePicker();
			this.enddate = new System.Windows.Forms.DateTimePicker();

			this.panel1.SuspendLayout();
			this.panel2.SuspendLayout();
			// 
			// panel1
			// 
			this.panel1.Controls.AddRange(new System.Windows.Forms.Control[] {
				this.panel2,
				this.findButton,
				this.userid,
				this.label1,
			    this.label2,
			    this.label3,
				this.viewtype,
				this.pagenumber,
				this.startdate,
				this.enddate });
			this.panel1.Name = "InputPanel";
			this.panel1.Size = new System.Drawing.Size(488, 248);
			this.panel1.TabIndex = 0;
			
			// 
			// label1
			// 
			this.label1.Location = new System.Drawing.Point(16, 10);
			this.label1.Name = "label1";
			this.label1.Size = new System.Drawing.Size(46, 16);
			this.label1.Text = "User";

			// 
			// userid
			// 
			this.userid.Location = new System.Drawing.Point(50, 6);
			this.userid.Name = "userid";
			this.userid.Size = new System.Drawing.Size(60, 20);
			this.userid.Text = "userid";
			this.userid.ReadOnly = true;

			// 
			// label2
			// 
			this.label2.Location = new System.Drawing.Point(118, 10);
			this.label2.Name = "label1";
			this.label2.Size = new System.Drawing.Size(32, 16);
			this.label2.Text = "View";

			
			// 
			// viewtype
			// 
			this.viewtype.Items.Add("Period");
			this.viewtype.Items.Add("Invoice");
			this.viewtype.SelectedIndex = 0;
			this.viewtype.Location = new System.Drawing.Point(148, 8);
			this.viewtype.Name = "viewtype";
			this.viewtype.Size = new System.Drawing.Size(100, 20);
			this.viewtype.TabIndex = 1;

			// 
			// label3
			// 
			this.label3.Location = new System.Drawing.Point(262, 10);
			this.label3.Name = "label1";
			this.label3.Size = new System.Drawing.Size(36, 16);
			this.label3.Text = "Page";

			// 
			// pagenumber
			// 
			this.pagenumber.Items.Add("1");
			this.pagenumber.SelectedIndex = 0;
			this.pagenumber.Location = new System.Drawing.Point(298, 8);
			this.pagenumber.Name = "pagenumber";
			this.pagenumber.Size = new System.Drawing.Size(60, 20);
			this.pagenumber.TabIndex = 1;

			// 
			// startdate
			// 
			this.startdate.Location = new System.Drawing.Point(16, 40);
			this.startdate.Name = "Start";
			this.startdate.Size = new System.Drawing.Size(200, 20);

			// 
			// enddate
			// 
			this.enddate.Location = new System.Drawing.Point(245, 40);
			this.enddate.Name = "End";
			this.enddate.Size = new System.Drawing.Size(200, 20);

			// 
			// findButton
			// 
			this.findButton.Location = new System.Drawing.Point(370, 6);
			this.findButton.Name = "findButton";
			this.findButton.TabIndex = 3;
			this.findButton.Text = "Find";
			this.findButton.Click += new System.EventHandler(this.find_Click);

			// 
			// panel2
			// 
			this.panel2.Controls.AddRange(new System.Windows.Forms.Control[] {
				this.result});
			this.panel2.Location = new System.Drawing.Point(0, 80);
			this.panel2.Name = "ResultPanel";
			this.panel2.Size = new System.Drawing.Size(456, 150);
			this.panel2.TabIndex = 4;
			// 
			// result
			// 
			this.result.Location = new System.Drawing.Point(8, 0);
			this.result.Name = "result";
			this.result.ReadOnly = true;
			this.result.Size = new System.Drawing.Size(440, 140);
			this.result.TabIndex = 0;
			this.result.Text = "";
			this.result.Enabled = false;


			this.panel1.ResumeLayout(false);
			this.panel2.ResumeLayout(false);
		}

		[STAThread]
		static void Main() 
		{
			Application.Run(new AccountStatus());
		}
	}
}

File: Code/Chapter04/AccountStatus/perl/AccountStatus.pl
=head1 NAME

userinfo.pl - Demonstrates the GetUserCall API

=head1 DESCRIPTION

This package provides an example of how to get user information
using the GetUserCall API and the User class.

=head1 COPYRIGHT

Copyright (C) 2003 Ray Rischpater. All rights reserved.

This program is free software; you can redistribute and/or modify
it under the same terms as Perl itself as long as the copyright
attribution remains.

This file is provided without any warranty, express or
implied, including but not limited to fitness of purpose.

This file is part of the book "Developing Applications with
eBay" written by Ray Rischpater and available from
Apress, LP.

=cut;

require eBaySampleApplication;
use Win32::OLE;
use Tk;
use Tk::ROText;
use Date;
use Data::Dumper;

my $application = eBaySampleApplication->new;
my $mw = $application->{MAINWINDOW};

$userid = $application->{APISESSION}->{RequestUserId};
$page = "0";

# Create the three frames
my $rootFrame = $mw->Frame;
my $inputFrame = $rootFrame->Frame;
my $outputFrame = $rootFrame->Frame;

# Fill the input frame
my $label1 = $inputFrame->Label( -text => "eBay User ID" );
my $entry1 = $inputFrame->Label( -textvariable => \$userid,
							     -relief => "sunken" );
my $label2 = $inputFrame->Label( -text => "View by" );
my $viewtype = $inputFrame->Listbox( -height => 2 );
my @items = qw( Period Invoice );
foreach (@items)
{
  $viewtype->insert( 'end', $_ );
}
$viewtype->selectionSet(0);
my $entry2 = $inputFrame->Entry( -textvariable => \$viewby,
							     -width => 3 );
my $label3 = $inputFrame->Label( -text => "Page" );
my $pagenumber = $inputFrame->Scrolled( "Listbox", 
										-height => 2,
									    -scrollbars => "oe" );
$pagenumber->insert( 'end', "1" );
Tk::grid( $label1, $entry1,
		  $label2, $viewtype,
		  $label3, $pagenumber ) ;

my $label4 = $inputFrame->Label( -text => "Start" );
$startdate = $inputFrame->Date( -datefmt => '%2m/%2d/%4y',
								-fields => 'date',
								-value => 'now' );
my $label5 = $inputFrame->Label( -text => "End" );
$enddate = $inputFrame->Date( -datefmt => '%2m/%2d/%4y',
							  -fields => 'date',
							  -value => 'now' );
my $button = $inputFrame->Button( -text => "Find",
								   -command => \&find_Click );
Tk::grid( $label4, $startdate, $label5, $enddate, $button );
# Fill the output frame
my $resultText = $outputFrame->Scrolled( "ROText", 
										 -height => 6,
										 -width => 64,
									     -scrollbars => "oe" )->pack;

# Pack the frames
$inputFrame->pack;
$outputFrame->pack;

$application->Gui( $rootFrame );

$application->Main;

#
# Do the eBay thing
#
sub find_Click
{
  $application->Busy( 1 );

  my $api = Win32::OLE->new( "eBay.SDK.API.GetAccountCall" );
  $api->{CompatibilityLevel} = 319;
  $api->{ApiCall}->{ApiCallSession} = $application->{APISESSION};
  $api->{ApiCall}->{ErrorLevel} = 1;

  ($api->{PageNumber}) = $pagenumber->curselection();
  $api->{PageNumber} = 1 if !$api->{PageNumber};
  ($api->{ViewType}) = $viewtype->curselection();
  $api->{ViewType} = 0 if !$api->{ViewType};

  if ( $api->{ViewType} == 0 )
  {
	my $popt = Win32::OLE->new( "eBay.SDK.Model.Account.PeriodViewSettings" );
	my $dr = Win32::OLE->new( "eBay.SDK.API.DateRangeImpl" );
	$dr->{BeginDate} = $startdate->get( "%Y-%m-%d" );
	$dr->{EndDate} = $enddate->get( "%Y-%m-%d" );
	$popt->{DateRange} = $dr;
	$api->{PeriodViewOption} = $popt;
  }

  if ( $api->{ViewType} == 1 )
  {
	my $ivopt = Win32::OLE->new( "eBay.SDK.Model.Account.InvoiceViewSettings" );
	$ivopt->{InvoiceMonth} = $startdate->get( "%m" );
	$ivopt->{InvoiceYear} = $startdate->get( "%y" );
	$api->{InvoiceViewOption} = $ivopt;
  }

  my $theAccount = $api->GetAccount();
  my $result = Win32::OLE->LastError();

  # Show the results
  if( $theAccount )
  {
	$result = "Account: " . $theAccount->{Id} . "\n";
	if ( $theAccount->{InvoiceView} )
	{
	  $result .= "Amount Due:\t" . $theAccount->{InvoiceView}->{InvoiceBalance} . "\n";
	  $result .= "Invoice Date:\t" . $theAccount->{InvoiceView}->{InvoiceDate}}->Date("MM/dd/yyyy") . "\n";
	}
	if ( $theAccount->{PeriodView} )
	{

		$result .= "Last Invoice Date:\t" . 
		  $theAccount->{PeriodView}->{LastInvoiceDate}->Date("MM/dd/yyyy") . "\n";
		$result .= "Last Invoice Amount:\t" .
		  $theAccount->{PeriodView}->{LastInvoiceAmount} . "\n";
		$result .= "Last Payment Date:\t" .
		  $theAccount->{PeriodView}->{LastPaymentDate}->Date("MM/dd/yyyy") . "\n";
		$result .= "Last Payment Amount:\t" .
		  $theAccount->{PeriodView}->{LastAmountPaid} . "\n";
		$result .= "Current Balance:\t" .
		  $theAccount->{PeriodView}->{CurrentBalance} . "\n";
	}

	# Enumerate each of the account activities
	$result .= "\nActivity History\n";
	my $i;
	for ( $i = 0; $i < $theAccount->{Activities}->ItemCount(); $i++ )
	{
	  my $activity = $theAccount->{Activities}->ItemAt($i);
	  $result .= $activity->{Date}->Date("MM/dd/yyyy") . "\t";
	  $result .= $activity->{Id} . "\t";
	  $result .= $activity->{Credit} . "\t" if $activity->{Credit};
  	  $result .= "(" . $activity->{Debit} . ")" . "\t" if $activity->{Debit};
	  $result .= "\n";
	}

	# Update the pagenumber menu with the list pages
	$pagenumber->delete( 0, "end" );
	for ( $i = 1; $i <= $api->{TotalPages}; $i++ )
    {
	  $pagenumber->insert( "end", "$i" );
	}
	
	$resultText->delete( "1.0", "end" );
	$resultText->insert( "end", $result );
  }
  $application->Busy( 0 );
}
